### PR TITLE
GIthub Discussions and Discussion Comments Webhooks

### PR DESCRIPTION
**Pull request by elewis787**
#### Summary
This pr adds basic support for GitHub discussions and discussion comment events. 

#### Ticket Link
If this pull request addresses a Help Wanted ticket https://github.com/mattermost/mattermost-plugin-github/issues/480 

#### Notes 

I would love to hear thoughts on the template for both discussions and discussion comments. I did not add additional style or attempt to pull out a lot of information but I am happy to update the template as needed. 

I also struggled with the playwrite test. If helpful, I am happy to further debug them and ensure they are working. I tested this manually using smee.io for webhook forwarding to a local deployment. I replayed the webhooks while testing. 

test where done by integrated https://github.com/orgs/elewis787-org/discussions 

I have attached a few screenshots to help demonstrate the functionality 

Screenshot 2024-07-23 at 7 38 40 AM
Screenshot 2024-07-23 at 7 38 06 AM

### CODE REVIEW

**Comment by elewis787** [comment]
`/check-cla`

File: server/plugin/webhook.go [hunk 1]
```diff
@@ -1382,3 +1393,90 @@ func (p *Plugin) postReleaseEvent(event *github.ReleaseEvent) {
...
+			UserId:  p.BotUserID,
+			Type:    "custom_git_discussion",
+			Message: newDiscussionMessage,
+		}
+		repoName := strings.ToLower(repo.GetFullName())
+		discussionNumber := event.GetDiscussion().GetNumber()
+
+		post.AddProp(postPropGithubRepo, repoName)
+		post.AddProp(postPropGithubObjectID, discussionNumber)
+		post.AddProp(postPropGithubObjectType, "dicussion")
```
Comment (fmartingr) [inline server/plugin/webhook.go:1430]:
```suggestion
		post.AddProp(postPropGithubObjectType, "discussion")
```

File: server/plugin/webhook.go [hunk 2]
```diff
@@ -1382,3 +1393,90 @@ func (p *Plugin) postReleaseEvent(event *github.ReleaseEvent) {
...
+		post.ChannelId = sub.ChannelID
+		if err = p.client.Post.CreatePost(post); err != nil {
+			p.client.Log.Warn("Error webhook post", "post", post, "error", err.Error())
+		}
+	}
+}
+
+func (p *Plugin) postDicussionCommentEvent(event *github.DiscussionCommentEvent) {
```
Comment (fmartingr) [inline server/plugin/webhook.go:1438]:
```suggestion
func (p *Plugin) postDiscussionCommentEvent(event *github.DiscussionCommentEvent) {
```

File: server/plugin/webhook.go [hunk 3]
```diff
@@ -287,6 +288,16 @@ func (p *Plugin) handleWebhook(w http.ResponseWriter, r *http.Request) {
 		handler = func() {
 			p.postReleaseEvent(event)
 		}
+	case *github.DiscussionEvent:
+		repo = event.GetRepo()
+		handler = func() {
+			p.postDicussionEvent(event)
```
Comment (fmartingr) [inline server/plugin/webhook.go:294]:
```suggestion
			p.postDiscussionEvent(event)
```

File: server/plugin/webhook.go [hunk 4]
```diff
@@ -287,6 +288,16 @@ func (p *Plugin) handleWebhook(w http.ResponseWriter, r *http.Request) {
...
+		}
+	case *github.DiscussionCommentEvent:
+		repo = event.GetRepo()
+		handler = func() {
+			p.postDicussionCommentEvent(event)
```
Comment (fmartingr) [inline server/plugin/webhook.go:299]:
```suggestion
			p.postDiscussionCommentEvent(event)
```

File: server/plugin/webhook.go [hunk 5]
```diff
@@ -1382,3 +1393,90 @@ func (p *Plugin) postReleaseEvent(event *github.ReleaseEvent) {
 		}
 	}
 }
+
+func (p *Plugin) postDicussionEvent(event *github.DiscussionEvent) {
```
Comment (fmartingr) [inline server/plugin/webhook.go:1397]:
```suggestion
func (p *Plugin) postDiscussionEvent(event *github.DiscussionEvent) {
```

**Review by fmartingr — CHANGES_REQUESTED** [review]
Good work on the PR! No specific complaints but there are a few typos around.

Can you get some screenshots and add those to the PR summary?

File: Makefile [hunk 6]
```diff
@@ -158,7 +158,7 @@ apply:
 ## Install go tools
 install-go-tools:
 	@echo Installing go tools
-	$(GO) install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.51.1
+	$(GO) install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.59.1
```
[the remark this answers is not in the document]
Comment (elewis787) [reply Makefile:161]: I ran into an issue with golangci-lint 

```
can't run linter goanalysis_metalinter: goanalysis_metalinter: atomicalign: package "root" (isInitialPkg: true, needAnalyzeSource: true): runtime error: invalid memory address or nil pointer dereference
```

First work around was to pull the latest version and see if that fixed the issue. After the update to the version I did not see the above error. 

I am running `go version go1.22.4 darwin/arm64`

File: server/plugin/template.go [hunk 7]
```diff
@@ -414,6 +414,8 @@ Assignees: {{range $i, $el := .Assignees -}} {{- if $i}}, {{end}}{{template "use
 		"    	* `pull_reviews` - includes pull request reviews\n" +
 		"    	* `releases` - includes release created and deleted\n" +
 		"    	* `label:<labelname>` - limit pull request and issue events to only this label. Must include `pulls` or `issues` in feature list when using a label.\n" +
+		"		* `discussions` - includes new discussions\n" +
```
[the remark this answers is not in the document]
Comment (elewis787) [reply server/plugin/template.go]: looks like 4 spaces and a tab is what the other lines used. I updated too match from 2 tabs.

**Comment by elewis787** [comment]
> Good work on the PR! No specific complaints but there are a few typos around.
> 
> Can you get some screenshots and add those to the PR summary?

Thanks @fmartingr. 

I have attached screenshots and a demo org/repository.

**Review by fmartingr — APPROVED** [review]
Thanks for the submission! I will request someone else's reviews to have more eyes on this and make sure it adheres to the plugin guidelines.

**Comment by fmartingr** [comment]
@elewis787 Can you check the lint errors?

**Comment by elewis787** [comment]
> @elewis787 Can you check the lint errors?

I pushed a fix for the indentation error in the const block. I'm looking into if I can retrigger the ci lint check to verify I caught all of them. I am not seeing any other changes locally.

File: server/plugin/template.go [hunk 8]
```diff
@@ -440,6 +442,16 @@ It now has **{{.GetRepo.GetStargazersCount}}** stars.`))
 {{- if eq .GetAction "created" }} created a release {{template "release" .GetRelease}}
 {{- else if eq .GetAction "deleted" }} deleted a release {{template "release" .GetRelease}}
 {{- end -}}`))
+
+	template.Must(masterTemplate.New("newDiscussion").Funcs(funcMap).Parse(
+		`Discussion [#{{.GetNumber}} {{.GetTitle}}]({{.GetHTMLURL}})`,
```
Comment (raghavaggarwal2308) [inline server/plugin/template.go:447]: Can we make this little more descriptive, like:

`@user stater a new discussion #number title on [org/repo]`

File: server/plugin/webhook.go [hunk 9]
```diff
@@ -1382,3 +1392,90 @@ func (p *Plugin) postReleaseEvent(event *github.ReleaseEvent) {
...
+
+		if p.excludeConfigOrgMember(event.GetSender(), sub) {
+			continue
+		}
+
+		post := &model.Post{
+			UserId:  p.BotUserID,
+			Type:    "custom_git_discussion",
+			Message: newDiscussionMessage,
+		}
```
Comment (raghavaggarwal2308) [inline server/plugin/webhook.go:1423]: you can use the `makeBotPost` function here, as you used in the func below

File: server/plugin/webhook.go [hunk 10]
```diff
@@ -1382,3 +1392,90 @@ func (p *Plugin) postReleaseEvent(event *github.ReleaseEvent) {
...
+		repoName := strings.ToLower(repo.GetFullName())
+		discussionNumber := event.GetDiscussion().GetNumber()
+
+		post.AddProp(postPropGithubRepo, repoName)
+		post.AddProp(postPropGithubObjectID, discussionNumber)
+		post.AddProp(postPropGithubObjectType, "discussion")
+		post.ChannelId = sub.ChannelID
+		if err = p.client.Post.CreatePost(post); err != nil {
+			p.client.Log.Warn("Error webhook post", "post", post, "error", err.Error())
```
Comment (raghavaggarwal2308) [inline server/plugin/webhook.go:1432]:
```suggestion
			p.client.Log.Warn("Error creating discussion notification post", "Post", post, "Error", err.Error())
```

File: server/plugin/webhook.go [hunk 11]
```diff
@@ -1382,3 +1392,90 @@ func (p *Plugin) postReleaseEvent(event *github.ReleaseEvent) {
...
+		repoName := strings.ToLower(repo.GetFullName())
+		commentID := event.GetComment().GetID()
+
+		post.AddProp(postPropGithubRepo, repoName)
+		post.AddProp(postPropGithubObjectID, commentID)
+		post.AddProp(postPropGithubObjectType, githubObjectTypeDiscussionComment)
+
+		if event.GetAction() == actionCreated {
+			post.Message = newDiscussionCommentMessage
+		}
```
Comment (raghavaggarwal2308) [inline server/plugin/webhook.go:1474]: We can remove this and directly pass the message int `createBotPost` func. We have already added the conditional for `actionCreated` on line 1445

File: server/plugin/webhook.go [hunk 12]
```diff
@@ -1382,3 +1392,90 @@ func (p *Plugin) postReleaseEvent(event *github.ReleaseEvent) {
...
+
+		post.ChannelId = sub.ChannelID
+		if err = p.client.Post.CreatePost(post); err != nil {
+			p.client.Log.Warn("Error webhook post", "post", post, "error", err.Error())
```
Comment (raghavaggarwal2308) [inline server/plugin/webhook.go:1478]:
```suggestion
			p.client.Log.Warn("Error creating discussion comment post", "Post", post, "Error", err.Error())
```

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]
@elewis787 Great work on this PR. Just added a few code refactoring comments

**Comment by fmartingr** [comment]
> @fmartingr There are already four other people who review the PR. If you don't mind, I prefer not to review the PR. Is that fine with you?

Yes! I just added a bunch of people because it's summer and I didn't knew who was available, we are enough as it is, thank you!

**Comment by elewis787** [comment]
> @elewis787 Great work on this PR. Just added a few code refactoring comments

Thanks! I will work through these today!

File: server/plugin/webhook.go (hunk 9 above)
Comment (elewis787) [reply server/plugin/webhook.go]: updated in commit 3a637da

File: server/plugin/template.go (hunk 8 above)
Comment (elewis787) [reply server/plugin/template.go]: updated both templates in commit 3a637da

File: package-lock.json
[the remark this answers is not in the document]
Comment (elewis787) [reply package-lock.json:1]: removed - 3a637da

File: webapp/package-lock.json
[the remark this answers is not in the document]
Comment (elewis787) [reply webapp/package-lock.json:1]: reverted back to master in commit 3a637da

**Comment by elewis787** [comment]
Thank you @raghavaggarwal2308 for the review! I believe I pushed the changes requested.

**Comment by elewis787** [comment]
Adding screen shots for the latest template updates. 

Screenshot 2024-07-29 at 10 27 10 AM
Screenshot 2024-07-29 at 10 27 18 AM

**Review by raghavaggarwal2308 — APPROVED** [review]
@elewis787 Thanks for you contribution, the changes looks good to me

**Comment by raghavaggarwal2308** [comment]
@wiggin77 Can you please give this PR a review. Also, I have never worked on playwright tests,so not sure about the changes make in the playwright tests.

**Comment by elewis787** [comment]
> @wiggin77 Can you please give this PR a review. Also, I have never worked on playwright tests,so not sure about the changes make in the playwright tests.

I see the package-lock.json shows edits - I can revert this as needed - these changes are a result of using an audit fix. I was not able to use the e2e test locally. I suspect that this is largely related to my local node version. I did not attempt to revert back to previous versions.

File: server/plugin/command.go [hunk 13]
```diff
@@ -15,37 +15,41 @@ import (
...
+	featurePullsMerged        = "pulls_merged"
+	featurePullsCreated       = "pulls_created"
+	featurePushes             = "pushes"
+	featureCreates            = "creates"
+	featureDeletes            = "deletes"
+	featureIssueComments      = "issue_comments"
+	featurePullReviews        = "pull_reviews"
+	featureStars              = "stars"
+	featureReleases           = "releases"
+	featureDisccusions        = "discussions"
```
Comment (wiggin77) [inline server/plugin/command.go:30]: Typo in variable name.

File: server/plugin/command.go [hunk 14]
```diff
@@ -15,37 +15,41 @@ import (
...
+	featurePullsMerged:        true,
+	featurePullsCreated:       true,
+	featurePushes:             true,
+	featureCreates:            true,
+	featureDeletes:            true,
+	featureIssueComments:      true,
+	featurePullReviews:        true,
+	featureStars:              true,
+	featureReleases:           true,
+	featureDisccusions:        true,
```
Comment (wiggin77) [inline server/plugin/command.go:51]: Typo in variable name here as well.

File: server/plugin/command.go (hunk 13 above)
Comment (elewis787) [reply server/plugin/command.go]: [95e2155](https://github.com/mattermost/mattermost-plugin-github/pull/808/commits/95e2155d254a1fc6ed29cb819f0cd511ddd0a0c5)

File: server/plugin/command.go (hunk 14 above)
Comment (elewis787) [reply server/plugin/command.go:51]: Updated in [95e2155](https://github.com/mattermost/mattermost-plugin-github/pull/808/commits/95e2155d254a1fc6ed29cb819f0cd511ddd0a0c5)

File: server/plugin/command.go [hunk 15]
```diff
@@ -15,37 +15,41 @@ import (
...
+	featurePullsMerged        = "pulls_merged"
+	featurePullsCreated       = "pulls_created"
+	featurePushes             = "pushes"
+	featureCreates            = "creates"
+	featureDeletes            = "deletes"
+	featureIssueComments      = "issue_comments"
+	featurePullReviews        = "pull_reviews"
+	featureStars              = "stars"
+	featureReleases           = "releases"
+	featureDiscusions         = "discussions"
```
Comment (wiggin77) [inline server/plugin/command.go:30]:
```suggestion
	featureDiscussions         = "discussions"
```

File: server/plugin/command.go (hunk 14 above)
Comment (wiggin77) [reply server/plugin/command.go:51]: @elewis787 still has a typo

```suggestion
	featureDiscussions:        true,
```

Comment (elewis787) [reply server/plugin/command.go:51]: Sorry about that - I committed the suggestion and updated the subscription check to use the updated spelling.

**Comment by raghavaggarwal2308** [comment]
> > @wiggin77 Can you please give this PR a review. Also, I have never worked on playwright tests,so not sure about the changes make in the playwright tests.
> 
> I see the package-lock.json shows edits - I can revert this as needed - these changes are a result of using an audit fix. I was not able to use the e2e test locally. I suspect that this is largely related to my local node version. I did not attempt to revert back to previous versions.

@elewis787 Yes, I think we can remove them

**Comment by elewis787** [comment]
> > > @wiggin77 Can you please give this PR a review. Also, I have never worked on playwright tests,so not sure about the changes make in the playwright tests.
> > 
> > 
> > I see the package-lock.json shows edits - I can revert this as needed - these changes are a result of using an audit fix. I was not able to use the e2e test locally. I suspect that this is largely related to my local node version. I did not attempt to revert back to previous versions.
> 
> @elewis787 Yes, I think we can revert that

I have push a reverted package-lock.json for the e2e tests.

**Review by wiggin77 — APPROVED** [review]
Great work @elewis787 :+1:

**Comment by elewis787** [comment]
Anything I can do to help push this forward?

**Comment by raghavaggarwal2308** [comment]
> Anything I can do to help push this forward?

Hey @elewis787 ,
This PR is waiting for a approval from QA. We will try to get it reviewed ASAP. 
Apologies for the delay.

**Comment by elewis787** [comment]
> > Anything I can do to help push this forward?
> 
> Hey @elewis787 , This PR is waiting for a approval from QA. We will try to get it reviewed ASAP. Apologies for the delay. cc: @AayushChaudhary0001

Thank! No rush. Please reach out if there is anything I can do to help.

**Review by AayushChaudhary0001 — APPROVED** [review]
This PR has been tested and looks good. All the notifications regarding the new discussion creation and discussion comments are received on Mattermost. Approved.

**Note:-** Please add the change regarding the events(`discussions`, `discussion comments`) to be enabled in Github webhook for notifications of discussions and discussion comments in the readme file.

**Comment by raghavaggarwal2308** [comment]
@elewis787 Can you create a PR in the MM docs https://github.com/mattermost/docs repo for he documentation changes? I think the webhook configurations needed to be updated for this PR to work.

**Comment by elewis787** [comment]
Absolutely! I'll work on a write up.

**Comment by raghavaggarwal2308** [comment]
@elewis787 Thanks! Please let us know when the PR is created. We can merge this PR after that

**Comment by elewis787** [comment]
Here are the minimal doc updates https://github.com/mattermost/docs/pull/7371. 

Please let me know if more details are needed.

**Comment by raghavaggarwal2308** [comment]
@elewis787 Awesome work on this PR. Thank you for your contribution.

**Comment by elewis787** [comment]
@raghavaggarwal2308 Thanks for all of your help with the reviews and direction! I am excited to see this merged.

**Pull request merged**